Sources/X509/TinyArray.swift

```diff
@@ -0,0 +1,230 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftCertificates open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftCertificates project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftCertificates project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+
+/// ``TinyArray`` is optimized to store zero or one ``Element``.
+/// It supports arbitrary many elements but if only up to one ``Element`` is stored it does **not** allocate seperate storage on the heap and
+/// instead stores the ``Element`` inline.
+@usableFromInline
+enum TinyArray<Element>: RandomAccessCollection {
```

Comment:
I don't love that callers can see it's an `enum` which to me is an implementation detail. That said it's internal so I'm also okay to leave it.